Blocking query of the health API "/health/service/<service>" may return empty nodes for non-empty services.

#### Overview of the Issue

For example, a GET call of the API `/health/service/consul?index=9999999&wait=10s` may return `status_code=200 index=1 nodes=[]` when the following event 2 happens immediately after event 1:
1. The consul token used in the request is revoked (for example, by the vault consul secret engine).
2. 10 seconds deadline reached,

This is due to a race condition of handling errors caused by the above two events:
1. Error handling of the first event is at [here](https://github.com/hashicorp/consul/blob/f461f49ea572968c04f2422a77ff05496f5e1e17/agent/submatview/rpc_materializer.go#L85-L87).  In line 86, `m.mat.reset()` resets the index.
2. Error handling of the second event is at [here](https://github.com/hashicorp/consul/blob/f461f49ea572968c04f2422a77ff05496f5e1e17/agent/submatview/materializer.go#L129-L135).  In line 133, `m.index=0` due to the above error handling.  So the `result.Value` is set to an empty list.

---

#### Reproduction Steps

1. Create a consul config file `consul.json`:
```json
{
  "server": true,
  "bootstrap" : true,
  "datacenter": "dc1",
  "data_dir": "/tmp/consul",
  "client_addr": "127.0.0.1",
  "advertise_addr": "127.0.0.1",
  "acl": {
    "enabled": true,
    "default_policy": "deny"
  }
}
```
2. Create a python script `reproduce.py`:
```python
import os
import json
import time
import subprocess
from multiprocessing import Process

import requests

CONSUL_HTTP_ADDR = os.getenv("CONSUL_HTTP_ADDR")

def create_token():
    cmd = "consul acl token create -policy-name=global-management -format=json"
    out = subprocess.check_output(cmd.split())
    return json.loads(out)

def delete_token(token):
    cmd = f"consul acl token delete -accessor-id={token['AccessorID']}"
    subprocess.check_call(cmd.split())

def get_consul_nodes(token, wait):
    url = CONSUL_HTTP_ADDR + "/v1/health/service/consul"
    url += f"?index=99999999&wait={wait}"
    headers = {
        "x-consul-token": token["SecretID"],
    }

    ret = requests.get(url, headers=headers)
    if ret.status_code == 200:
        nodes = ret.json()
        if len(nodes) == 0:
            print(f"=== status_code={ret.status_code} index={ret.headers['X-Consul-Index']} nodes={nodes}")

def main():
    token = create_token()
    delays = range(10)
    blocking_wait_seconds = 10

    def f(d):
        time.sleep(d/len(delays))
        get_consul_nodes(token, f"{blocking_wait_seconds}s")

    procs = []
    for d in delays:
        p = Process(target=f, args=(d,))
        p.start()
        procs.append(p)

    time.sleep(blocking_wait_seconds)
    delete_token(token)

    for p in procs:
        p.join()
```
3. Start consul server
```
$ consul agent -config-file consul.json
```
4. Run the reproduce script:
```
$ consul acl bootstrap
...
SecretID:         xxx
Policies:
   00000000-0000-0000-0000-000000000001 - global-management
...
$ export CONSUL_HTTP_ADDR=http://127.0.0.1:8500
$ export CONSUL_HTTP_TOKEN=xxx
$ python reproduce.py
=== status_code=200 index=1 nodes=[]
=== status_code=200 index=1 nodes=[]
=== status_code=200 index=1 nodes=[]
=== status_code=200 index=1 nodes=[]
=== status_code=200 index=1 nodes=[]
=== status_code=200 index=1 nodes=[]
=== status_code=200 index=1 nodes=[]
=== status_code=200 index=1 nodes=[]
=== status_code=200 index=1 nodes=[]
```

### Consul info

Reproduced in consul 1.11.7.  Also reproduced in the main branch commit: 6f48ce153c9

### Operating system and Environment details

Arch Linux Kernel v6.6.6

### Log Fragments
```
2024-03-01T15:37:06.262-0800 [ERROR] agent.rpcclient.health: subscribe call failed: err="stream reset requested" failure_count=1 key=consul m=0xc001761ae0 topic=ServiceHealth
2024-03-01T15:37:12.451-0800 [ERROR] agent.rpcclient.health: subscribe call failed: err="rpc error: code = Unknown desc = ACL not found" failure_count=2 key=consul m=0xc001761ae0 topic=ServiceHealth
```

--- a/.changelog/20876.txt
+++ b/.changelog/20876.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+streaming: Handle ACL errors consistently when blocking query timeout is reached.
+```

--- a/agent/consul/state/store_integration_test.go
+++ b/agent/consul/state/store_integration_test.go
@@ -72,7 +72,7 @@ func TestStore_IntegrationWithEventPublisher_ACLTokenUpdate(t *testing.T) {
 
 	// Ensure the reset event was sent.
 	err = assertErr(t, eventCh)
-	require.Equal(t, stream.ErrSubForceClosed, err)
+	require.Equal(t, stream.ErrACLChanged, err)
 
 	// Register another subscription.
 	subscription2 := &stream.SubscribeRequest{
@@ -101,7 +101,7 @@ func TestStore_IntegrationWithEventPublisher_ACLTokenUpdate(t *testing.T) {
 
 	// Ensure the reset event was sent.
 	err = assertErr(t, eventCh2)
-	require.Equal(t, stream.ErrSubForceClosed, err)
+	require.Equal(t, stream.ErrACLChanged, err)
 }
 
 func TestStore_IntegrationWithEventPublisher_ACLPolicyUpdate(t *testing.T) {
@@ -191,7 +191,7 @@ func TestStore_IntegrationWithEventPublisher_ACLPolicyUpdate(t *testing.T) {
 
 	// Ensure the reload event was sent.
 	err = assertErr(t, eventCh)
-	require.Equal(t, stream.ErrSubForceClosed, err)
+	require.Equal(t, stream.ErrACLChanged, err)
 
 	// Register another subscription.
 	subscription3 := &stream.SubscribeRequest{
@@ -381,7 +381,7 @@ func assertReset(t *testing.T, eventCh <-chan nextResult, allowEOS bool) {
 				}
 			}
 			require.Error(t, next.Err)
-			require.Equal(t, stream.ErrSubForceClosed, next.Err)
+			require.Equal(t, stream.ErrACLChanged, next.Err)
 			return
 		case <-time.After(100 * time.Millisecond):
 			t.Fatalf("no err after 100ms")

--- a/agent/consul/stream/event_publisher.go
+++ b/agent/consul/stream/event_publisher.go
@@ -393,7 +393,7 @@ func (s *subscriptions) closeSubscriptionsForTokens(tokenSecretIDs []string) {
 	for _, secretID := range tokenSecretIDs {
 		if subs, ok := s.byToken[secretID]; ok {
 			for _, sub := range subs {
-				sub.forceClose()
+				sub.closeACLChanged()
 			}
 		}
 	}

--- a/agent/consul/stream/subscription.go
+++ b/agent/consul/stream/subscription.go
@@ -25,9 +25,13 @@ const (
 	// will not return new events.
 	subStateUnsub = 2
 
-	// subStateShutting down indicates the subscription was closed due to
+	// subStateShuttingDown indicates the subscription was closed due to
 	// the server being shut down.
 	subStateShuttingDown = 3
+
+	// subStateACLChanged indicates the subscription was closed due to
+	// a change in ACLs.
+	subStateACLChanged = 4
 )
 
 // ErrSubForceClosed is a error signalling the subscription has been
@@ -38,6 +42,12 @@ var ErrSubForceClosed = errors.New("subscription closed by server, client must r
 // been closed because the server is shutting down. The client should
 // subscribe to a different server to get streaming event updates.
 var ErrShuttingDown = errors.New("subscription closed by server, server is shutting down")
+
+// ErrACLChanged is an error to signal that the subscription has
+// been closed because a change in ACL token or its associated roles or policies has occurred.
+// If the token or policy is no longer valid, the client should resubscribe using a valid token. Otherwise,
+// the client should resubscribe using the same token.
+var ErrACLChanged = errors.New("subscription closed by server, ACL change occurred")
 
 // Subscription provides events on a Topic. Events may be filtered by Key.
 // Events are returned by Next(), and may start with a Snapshot of events.
@@ -131,6 +141,8 @@ func (s *Subscription) requireStateOpen() error {
 		return ErrSubForceClosed
 	case subStateShuttingDown:
 		return ErrShuttingDown
+	case subStateACLChanged:
+		return ErrACLChanged
 	case subStateUnsub:
 		return fmt.Errorf("subscription was closed by unsubscribe")
 	default:
@@ -162,6 +174,14 @@ func (s *Subscription) forceClose() {
 // Close the subscription and indicate that the server is being shut down.
 func (s *Subscription) shutDown() {
 	if atomic.CompareAndSwapUint32(&s.state, subStateOpen, subStateShuttingDown) {
+		close(s.closed)
+	}
+}
+
+// Close the subscription and indicate that an ACL change occurred. This change may require
+// a client to subscribe with a new token or re-subscribe with an existing token.
+func (s *Subscription) closeACLChanged() {
+	if atomic.CompareAndSwapUint32(&s.state, subStateOpen, subStateACLChanged) {
 		close(s.closed)
 	}
 }

--- a/agent/grpc-external/services/connectca/watch_roots.go
+++ b/agent/grpc-external/services/connectca/watch_roots.go
@@ -47,7 +47,7 @@ func (s *Server) WatchRoots(_ *pbconnectca.WatchRootsRequest, serverStream pbcon
 	for {
 		var err error
 		idx, err = s.serveRoots(options.Token, idx, serverStream, logger)
-		if errors.Is(err, stream.ErrSubForceClosed) {
+		if errors.Is(err, stream.ErrSubForceClosed) || errors.Is(err, stream.ErrACLChanged) {
 			logger.Trace("subscription force-closed due to an ACL change or snapshot restore, will attempt to re-auth and resume")
 		} else {
 			return err
@@ -90,7 +90,7 @@ func (s *Server) serveRoots(
 	for {
 		event, err := sub.Next(serverStream.Context())
 		switch {
-		case errors.Is(err, stream.ErrSubForceClosed):
+		case errors.Is(err, stream.ErrSubForceClosed) || errors.Is(err, stream.ErrACLChanged):
 			// If the subscription was closed because the state store was abandoned (e.g.
 			// following a snapshot restore) reset idx to ensure we don't skip over the
 			// new store's events.

--- a/agent/grpc-external/services/serverdiscovery/watch_servers.go
+++ b/agent/grpc-external/services/serverdiscovery/watch_servers.go
@@ -40,7 +40,7 @@ func (s *Server) WatchServers(req *pbserverdiscovery.WatchServersRequest, server
 	for {
 		var err error
 		idx, err = s.serveReadyServers(options.Token, idx, req, serverStream, logger)
-		if errors.Is(err, stream.ErrSubForceClosed) {
+		if errors.Is(err, stream.ErrSubForceClosed) || errors.Is(err, stream.ErrACLChanged) {
 			logger.Trace("subscription force-closed due to an ACL change or snapshot restore, will attempt to re-auth and resume")
 		} else {
 			return err
@@ -69,7 +69,7 @@ func (s *Server) serveReadyServers(token string, index uint64, req *pbserverdisc
 	for {
 		event, err := sub.Next(serverStream.Context())
 		switch {
-		case errors.Is(err, stream.ErrSubForceClosed):
+		case errors.Is(err, stream.ErrSubForceClosed) || errors.Is(err, stream.ErrACLChanged):
 			return index, err
 		case errors.Is(err, context.Canceled):
 			return 0, nil

--- a/agent/grpc-internal/services/subscribe/subscribe.go
+++ b/agent/grpc-internal/services/subscribe/subscribe.go
@@ -79,8 +79,7 @@ func (h *Server) Subscribe(req *pbsubscribe.SubscribeRequest, serverStream pbsub
 			return status.Error(codes.Aborted, err.Error())
 		case errors.Is(err, stream.ErrACLChanged):
 			logger.Trace("ACL change occurred; re-authenticating")
-			var authzErr error
-			authz, authzErr = h.Backend.ResolveTokenAndDefaultMeta(req.Token, &entMeta, nil)
+			_, authzErr := h.Backend.ResolveTokenAndDefaultMeta(req.Token, &entMeta, nil)
 			if authzErr != nil {
 				return authzErr
 			}

--- a/agent/grpc-internal/services/subscribe/subscribe.go
+++ b/agent/grpc-internal/services/subscribe/subscribe.go
@@ -79,10 +79,13 @@ func (h *Server) Subscribe(req *pbsubscribe.SubscribeRequest, serverStream pbsub
 			return status.Error(codes.Aborted, err.Error())
 		case errors.Is(err, stream.ErrACLChanged):
 			logger.Trace("ACL change occurred; re-authenticating")
-			authz, err = h.Backend.ResolveTokenAndDefaultMeta(req.Token, &entMeta, nil)
-			if err != nil {
-				return err
+			var authzErr error
+			authz, authzErr = h.Backend.ResolveTokenAndDefaultMeta(req.Token, &entMeta, nil)
+			if authzErr != nil {
+				return authzErr
 			}
+			// Otherwise, abort the stream so the client re-subscribes.
+			return status.Error(codes.Aborted, err.Error())
 		case err != nil:
 			return err
 		}

--- a/agent/grpc-internal/services/subscribe/subscribe.go
+++ b/agent/grpc-internal/services/subscribe/subscribe.go
@@ -77,6 +77,12 @@ func (h *Server) Subscribe(req *pbsubscribe.SubscribeRequest, serverStream pbsub
 		case errors.Is(err, stream.ErrSubForceClosed):
 			logger.Trace("subscription reset by server")
 			return status.Error(codes.Aborted, err.Error())
+		case errors.Is(err, stream.ErrACLChanged):
+			logger.Trace("ACL change occurred; re-authenticating")
+			authz, err = h.Backend.ResolveTokenAndDefaultMeta(req.Token, &entMeta, nil)
+			if err != nil {
+				return err
+			}
 		case err != nil:
 			return err
 		}

--- a/agent/grpc-internal/services/subscribe/subscribe_test.go
+++ b/agent/grpc-internal/services/subscribe/subscribe_test.go
@@ -992,28 +992,28 @@ node "node1" {
 		}
 	})
 
-	testutil.RunStep(t, "invalid token should return an error", func(t *testing.T) {
-		// Re-subscribe because the previous test step terminated the stream.
-		chEvents = make(chan eventOrError, 0)
-		streamClient := pbsubscribe.NewStateChangeSubscriptionClient(conn)
-		streamHandle, err := streamClient.Subscribe(ctx, &pbsubscribe.SubscribeRequest{
-			Topic: pbsubscribe.Topic_ServiceHealth,
-			Subject: &pbsubscribe.SubscribeRequest_NamedSubject{
-				NamedSubject: &pbsubscribe.NamedSubject{
-					Key: "foo",
-				},
+	// Re-subscribe because the previous test step terminated the stream.
+	chEvents = make(chan eventOrError, 0)
+	streamClient := pbsubscribe.NewStateChangeSubscriptionClient(conn)
+	streamHandle, err := streamClient.Subscribe(ctx, &pbsubscribe.SubscribeRequest{
+		Topic: pbsubscribe.Topic_ServiceHealth,
+		Subject: &pbsubscribe.SubscribeRequest_NamedSubject{
+			NamedSubject: &pbsubscribe.NamedSubject{
+				Key: "foo",
 			},
-			Token: token,
-		})
-		require.NoError(t, err)
+		},
+		Token: token,
+	})
+	require.NoError(t, err)
 
-		go recvEvents(chEvents, streamHandle)
+	go recvEvents(chEvents, streamHandle)
 
-		// Stub out token authn function so that the token is no longer considered valid.
-		backend.resolveTokenAndDefaultMeta = func(t string, entMeta *acl.EnterpriseMeta, _ *acl.AuthorizerContext) (acl.Authorizer, error) {
-			return nil, fmt.Errorf("ACL not found")
-		}
+	// Stub out token authn function so that the token is no longer considered valid.
+	backend.resolveTokenAndDefaultMeta = func(t string, entMeta *acl.EnterpriseMeta, _ *acl.AuthorizerContext) (acl.Authorizer, error) {
+		return nil, fmt.Errorf("ACL not found")
+	}
 
+	testutil.RunStep(t, "invalid token should return an error", func(t *testing.T) {
 		// Force another ACL update.
 		tokenID, err := uuid.GenerateUUID()
 		require.NoError(t, err)

--- a/agent/grpc-internal/services/subscribe/subscribe_test.go
+++ b/agent/grpc-internal/services/subscribe/subscribe_test.go
@@ -6,6 +6,7 @@ package subscribe
 import (
 	"context"
 	"errors"
+	"fmt"
 	"io"
 	"net"
 	"testing"
@@ -323,17 +324,21 @@ func getEvent(t *testing.T, ch chan eventOrError) *pbsubscribe.Event {
 }
 
 type testBackend struct {
-	publisher   *stream.EventPublisher
-	store       *state.Store
-	authorizer  func(token string, entMeta *acl.EnterpriseMeta) acl.Authorizer
-	forwardConn *gogrpc.ClientConn
+	publisher                  *stream.EventPublisher
+	store                      *state.Store
+	authorizer                 func(token string, entMeta *acl.EnterpriseMeta) acl.Authorizer
+	resolveTokenAndDefaultMeta func(token string, entMeta *acl.EnterpriseMeta, _ *acl.AuthorizerContext) (acl.Authorizer, error)
+	forwardConn                *gogrpc.ClientConn
 }
 
 func (b testBackend) ResolveTokenAndDefaultMeta(
 	token string,
 	entMeta *acl.EnterpriseMeta,
-	_ *acl.AuthorizerContext,
+	authCtx *acl.AuthorizerContext,
 ) (acl.Authorizer, error) {
+	if b.resolveTokenAndDefaultMeta != nil {
+		return b.resolveTokenAndDefaultMeta(token, entMeta, authCtx)
+	}
 	return b.authorizer(token, entMeta), nil
 }
 
@@ -984,6 +989,47 @@ node "node1" {
 			require.Equal(t, codes.Aborted, s.Code())
 		case <-time.After(2 * time.Second):
 			t.Fatalf("timeout waiting for aborted error")
+		}
+	})
+
+	testutil.RunStep(t, "invalid token should return an error", func(t *testing.T) {
+		// Re-subscribe because the previous test step terminated the stream.
+		chEvents = make(chan eventOrError, 0)
+		streamClient := pbsubscribe.NewStateChangeSubscriptionClient(conn)
+		streamHandle, err := streamClient.Subscribe(ctx, &pbsubscribe.SubscribeRequest{
+			Topic: pbsubscribe.Topic_ServiceHealth,
+			Subject: &pbsubscribe.SubscribeRequest_NamedSubject{
+				NamedSubject: &pbsubscribe.NamedSubject{
+					Key: "foo",
+				},
+			},
+			Token: token,
+		})
+		require.NoError(t, err)
+
+		go recvEvents(chEvents, streamHandle)
+
+		// Stub out token authn function so that the token is no longer considered valid.
+		backend.resolveTokenAndDefaultMeta = func(t string, entMeta *acl.EnterpriseMeta, _ *acl.AuthorizerContext) (acl.Authorizer, error) {
+			return nil, fmt.Errorf("ACL not found")
+		}
+
+		// Force another ACL update.
+		tokenID, err := uuid.GenerateUUID()
+		require.NoError(t, err)
+
+		aclToken := &structs.ACLToken{
+			AccessorID: tokenID,
+			SecretID:   token,
+		}
+		require.NoError(t, backend.store.ACLTokenSet(ids.Next("update"), aclToken))
+
+		select {
+		case item := <-chEvents:
+			require.Error(t, item.err, "got event instead of an error: %v", item.event)
+			require.EqualError(t, item.err, "rpc error: code = Unknown desc = ACL not found")
+		case <-time.After(2 * time.Second):
+			t.Fatalf("timeout waiting for ACL not found error")
 		}
 	})
 }

